core/src/bank_forks.rs

```diff
@@ -211,49 +336,123 @@ impl BankForks {
+
+        // Create the storage links to persist and track the necessary storage files
+        let account_storage_entries = bank.rc.get_storage_entries();
+        Self::add_snapshot_storage_links(slot, &slot_snapshot_dir, &account_storage_entries[..])
+            .map_err(Error::from)?;
+
         Ok(())
     }
 
-    pub fn remove_snapshot(slot: u64, path: &Option<String>) {
-        let path = BankForks::get_snapshot_path(path);
-        let bank_file = format!("{}", slot);
-        let bank_file_path = path.join(bank_file);
-        let _ = fs::remove_file(bank_file_path);
+    // 1) Create hard links to the AppendVec in each AccountStorageEntry that's necessary to restore from this
+    // snapshot, so that these AppendVec's don't get removed.
+    // 2) Create symlinks to each of the hard links to track them for deletion when
+    // the snapshot is deleted
+    fn add_snapshot_storage_links<P: AsRef<Path>, A: AsRef<AccountStorageEntry>>(
+        slot: u64,
+        snapshot_path: P,
+        account_storage_entries: &[A],
+    ) -> Result<()> {
+        let storage_soft_link_dir = Self::get_bank_snapshot_soft_link_dir(snapshot_path, slot);
+        fs::create_dir_all(&storage_soft_link_dir).map_err(Error::from)?;
+
+        for storage_entry in account_storage_entries {
+            let append_vec_file_path = &storage_entry.as_ref().get_path();
+
+            let account_storage_dir_path = append_vec_file_path
+                .parent()
+                .expect("Invalid AppendVec file path");
+
+            let link_file_name = Self::get_snapshot_storage_entry_link_name(
+                slot,
+                storage_entry.as_ref().fork_id(),
+                storage_entry.as_ref().append_vec_id(),
+            );
+
+            // Create the hard link alongside the AppendVec file
+            let hard_link = account_storage_dir_path.join(&link_file_name);
+            fs::hard_link(&append_vec_file_path, &hard_link)?;
+
+            // Track the newly created hardlink as a symlink inside the `storage_soft_links`
+            // directory for the hardlink so we can remove it when the snapshot is deleted
+            let soft_link = storage_soft_link_dir.join(&link_file_name);
+            symlink::symlink_file(&hard_link, &soft_link)?;
+        }
+
+        Ok(())
     }
 
-    pub fn set_snapshot_config(&mut self, path: Option<String>) {
-        self.snapshot_path = path;
+    pub fn remove_snapshot<P: AsRef<Path>>(slot: u64, snapshot_path: P) -> Result<()> {
+        let slot_snapshot_dir = Self::get_bank_snapshot_dir(&snapshot_path, slot);
+
+        // Remove the snapshot file
+        let bank_file_path = slot_snapshot_dir.join(slot.to_string());
+        fs::remove_file(bank_file_path)?;
+
+        let soft_link_dir = Self::get_bank_snapshot_soft_link_dir(&snapshot_path, slot);
+
+        // Remove all the hard links pointed to by the symlinks
+        for entry in fs::read_dir(&soft_link_dir)? {
```

Comment:
@rob-solana, this is where I'm deleting the hard links tracked by the sym links on snapshot deletion.